shrinkwrap: handle devDeps separately

bengl opened the pull request:
We've been running into issues caused by #5403, so this a PR to implement it. 

As a result of this PR, the following things happen:
- `devDependency`-only items in the shrinkwrap are marked with `dev: true`.
- `npm i --only=prod` will not install anything in the shrinkwrap marked `dev: true`.
- `npm i --only=dev` will only install things in the shrinkwrap marked with `dev: true`.

othiym23 commented:
This is promising, and I understand the use case for your team and think that this is the least disruptive way to handle it, so where is this at?

bengl commented:
I should have time to work on this some more in the next few days. Sorry for the delay!

bengl committed:
shrinkwrap: add dev property to devDeps

bengl committed:
install, shrinkwrap: respect dev/prod data in shrinkwrap file during install

terinjokes committed:
shrinkwrap: exclude dependencies of devDependencies

As the tree is now flattened, dependencies of devDependencies weren't
automatically excluded simply by not recursing into a devDependency's
children.

Instead, now explicitly walk "up" a dependency's tree to the root to
determine if it's a devDependency or solely a dependency of
a devDependency.

Resolve #9727

bengl commented:
#10265 seems super-helpful for this.

bengl committed:
test: install, shrinkwrap tests for new 'dev' property

bengl commented:
OK, by including @terinjokes's [PR](https://github.com/npm/npm/pull/10265) (thanks!), this became a whole lot easier, so here it is, including tests. I edited the title and description of this PR to show completion and include the net effects. :+1: :boom:

othiym23 commented:
Awesome! I'll talk to @iarna about including this in next week's release (it's a little late in the week to be including in tomorrow's, especially because we've been doing a lot of cheese-moving around shrinkwrap over the last few weeks, so things are maybe not quite as solid as we'd like, and we need to review these kinds of patches carefully). Thanks to both of you for putting this together! (We'll also probably land @terinjokes's change first and then rebase this patch on top of that, so plz don't squash any commits until we get a chance to do that.)

avindra commented:
Thanks for putting this together. This will allow us to do better CI in our project.

@andersjanmyr  : Take a look. P.S., have you found any workarounds in the meantime (other than merging your "dependencies" and "devDependencies" lists?

andersjanmyr commented:
@avindra We ended up just packaging everything up in a tar-file for production after a clean build
without the shrinkwrap. This has worked fine, since we are using the same architecture for development and for production. We mainly use shrinkwrap to pin down our development dependencies, especially for grunt-plugins.

terinjokes commented:
What is the expected behavior of shrinkwrapping optionalDependencies?

Right now optionalDependencies are included in the shrinkwrap file, and can cause an installation failure when they fail to hydrate. For example, when a shrinkwrap is created on an OS X machine, thus including "fsevents", and an install is then attempted on Linux.

Should we tag optionalDependencies, as this pull request does for devDependencies, and allow them to fail on hydration?

bengl commented:
Provided something like `isTreeOptional` can be made, sure. Sounds like it would solve #2679.

terinjokes commented:
@bengl I've got some free time tomorrow. I'll look into if it can be made.

evocateur commented:
@terinjokes Any progress on the `isTreeOptional` stuff? I'm feeling the pain with shrinkwrapped `fsevents` on a linux host, so I'm willing to help out however I can.

terinjokes commented:
@evocateur not yet, i'll pop it off the queue shortly.

Whoaa512 commented:
Also, what happens in these cases:

Prerequisites:
- Existing `npm-shrinkwrap.json` which includes both dev and prod deps

Cases:
- `npm install --save-dev new_package` - Will the `new_package` be added to the shrinkwrap?
- `npm install --save-dev existing_package@NEW.VERSION` - Will the `existing_package` be updated in the shrinkwrap?
- `npm install --save existing_package@NEW.VERSION` - Will the existing dev deps remain untouched in the shrinkwrap?

evocateur commented:
@Whoaa512 In all cases, you need to run `npm shrinkwrap` after the install commands, otherwise there will be no changes to `npm-shrinkwrap.json`.

Whoaa512 commented:
What about when `--also=dev` is added to the aforementioned cases? Looking at the install code and from local testing `saveShrinkwrap` inside install does not respect this setting, even though the docs/changelog would tell you otherwise.

bengl committed:
optionalDependencies get optional property in shrinkwrap

iarna commented:
@terinjokes I've verified that failing optional shrinkwrapped dependencies don't break the install, they just warn. I see this behavior going as far back `3.3.12`.

Now there have been some more complicated wrinkles around these, so I could believe there's a bug lurking, but a reproducer would be really helpful here!

bengl commented:
@iarna here's a contrived example:


package.json:
```js
{
  "optionalDependencies": {
    "test-c": "7.0.0"
  }
}
```

npm-shrinkwrap.json:
```js
{
  "dependencies": {
    "test-c": {
      "version": "7.0.0",
      "from": "test-c@7.0.0",
      "resolved": "https://registry.npmjs.org/test-c/-/test-c-7.0.0.tgz"
    }
  }
}
```

The optionalDependency is a version that's not on the registry, and so the shrinkwrap file could never really be generated, but I think this properly simulates a version having been unpublished.

* `npm install` with these two files fails, with fetch timeouts on the non-existent package.
* `npm install` without the shrinkwrap succeeds, with warning.

iarna commented:
Ok, I'm sorry I'm just really digging into the proposed solution now but…

> devDependency-only items in the shrinkwrap are marked with dev: true.
> npm i --only=prod will not install anything in the shrinkwrap marked dev: true.
> npm i --only=dev will only install things in the shrinkwrap marked with dev: true.

Why are we adding anything to the shrinkwrap? Why not do these checks in inflate-shrinkwrap?

bengl commented:
@iarna Hmm, it's been a while since I did this, but as I recall, it was my understanding that knowledge of dev-or-not was not available in inflate-shrinkwrap. Is it there in the tree object?

iarna commented:
It is! So somewhere in the asyncMap I'd…

https://github.com/npm/npm/blob/master/lib/install/inflate-shrinkwrap.js#L22

Add a check like:
`if (tree.package.devDependencies[name]) {`

If you're skipping it just `return next()`

bengl commented:
Oh awesome! That simplifies things. I'd still need `isTreeDev` (and the soon to arrive `isTreeOptional`) though, right? Otherwise we're only looking at one level.

iarna commented:
I was thinking that at first too, but no, because you never add that child to the tree, you never recurse to any of its children, and so there's never any need to check on them.

The only other detail, which is probably an outstanding bug, is:

https://github.com/npm/npm/blob/master/lib/install/deps.js#L110

which should account for `only=dev(elopment)` and `only=prod(uction)` too.

terinjokes commented:
@iarna when installing from shrinkwrap everything is (mostly) flat including dependencies of that devDependency.

iarna commented:
@terinjokes Ah yeah, fair, I was thinking the work flow based around resolving deps initially, but you're right, that doesn't apply here. Instead determining it at run time means walking up `requiredBy` and seeing if all of the ultimate top levels are dev deps.

Rush commented:
Is there a chance of merging this PR any time soon?

terinjokes commented:
@Rush I think they're waiting on me to update to the new style. I plan on getting to it before early next week.

graingert commented:
@terinjokes what's the status of this PR?

terinjokes commented:
@graingert I need to make changes to make npm happy, and I haven't had the time or energy to do finish.

jgrund commented:
What changes need to be made to get this to land? I'd really like to have this functionality.

iarna commented:
Ok, I'll restate what's remaining on this PR:

1. This will need to be rebased on to master.

2. Previously I'd asked that the use of `child.package._requiredBy` be rewritten to use a newer data structure `child.requiredBy`. Since then (as of 3.10.0) `_requiredBy` has gone away entirely and as such is no longer available to use at all. So this is no longer a "would make a better PR" and is now "the PR won't work without this change".

graingert commented:
ping @bengl

bengl commented:
@iarna I'm giving this another go this afternoon/evening to see if I can rebase and use `child.requiredBy`.

bengl committed:
[squash] make it work off current master

bengl commented:
OK, seems to be working as intended, after rebasing on master and making the required tweaks. I think this is good now, modulo squashing?

bengl commented:
Looks like an OOM failure on travis and appveyor seemed to have git perm issues? Either way, unrelated.

graingert commented:
@bengl I think every PR is doing this.

@bengl you need to rebase again.

graingert commented:
@othiym23 @iarna any chance of getting this into a release?

evocateur commented:
I rebased it locally to latest `master`, the only tricksy part was re-establishing the `dev` and `prod` checks in [`inflate-shrinkwrap.js`](https://github.com/npm/npm/compare/master...evocateur:devdepshrink-rebased#diff-1429e2e6780fcf5f0902e6ed7d4c3afbR27).

I would really, really like this to get into the next release.

graingert commented:
So say we all!

bengl commented:
Cool I rebased again. This is gonna need some changes when https://github.com/npm/npm/compare/iarna/fix-dev-prod-env is merged.

terinjokes commented:
@bengl thanks for getting this to move forward. let me know if there's anything I can do to help.

iarna commented:
`fix-dev-prod-env` is not a blocker on this (it's a minor structural change and will need tests to land). 

@bengl Where is this at here? Do you think this is ready to land?

bengl commented:
@iarna It passes the tests I had for it. I think it's good to go.

graingert commented:
@bengl @iarna any updates on this?

iarna commented:
I'm putting this on my list to finish and get landed.

graingert commented:
@iarna casual bump.

iarna commented:
Ok, the current status of this is:

This adds new attributes to `npm-shrinkwrap.json` for `optional` and `dev` to record what kind of dependency they were saved as.

It makes it so that `npm install --only=prod` _won't_ install deps that were saved to the shrinkwrap as dev deps.

Currently at adds no new behavior for the optional flagging.

iarna commented:
I'm ok with landing this as is and we can add behavior around optional deps later as desired.

othiym23 commented:
SGTM.

iarna commented:
I'm just running the tests for the ported and squashed version and if all looks good I'll land it.

iarna commented:
Looks good!

iarna commented:
Merged to our `release-next` branch. This will be in the next release of npm (scheduled for September 8th).
